File: tests/src/org.apache.hadoop.thirdparty/hadoop-shaded-protobuf_3_7/1.1.1/src/test/java/org_apache_hadoop_thirdparty/hadoop_shaded_protobuf_3_7/GeneratedMessageLiteInnerSerializedFormTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_hadoop_thirdparty.hadoop_shaded_protobuf_3_7;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.OutputStream;
import java.util.Arrays;

import org.apache.hadoop.thirdparty.protobuf.ByteString;
import org.apache.hadoop.thirdparty.protobuf.CodedInputStream;
import org.apache.hadoop.thirdparty.protobuf.CodedOutputStream;
import org.apache.hadoop.thirdparty.protobuf.ExtensionRegistryLite;
import org.apache.hadoop.thirdparty.protobuf.GeneratedMessageLite;
import org.apache.hadoop.thirdparty.protobuf.InvalidProtocolBufferException;
import org.apache.hadoop.thirdparty.protobuf.MessageLite;
import org.apache.hadoop.thirdparty.protobuf.Parser;
import org.junit.jupiter.api.Test;

public class GeneratedMessageLiteInnerSerializedFormTest {
    private static final byte[] MODERN_PAYLOAD = new byte[] {1, 3, 5, 7};
    private static final byte[] LEGACY_PAYLOAD = new byte[] {2, 4, 6, 8};

    @Test
    void readResolveRestoresMessageThroughDefaultInstanceField() throws Exception {
        Object restored = serializeFormAndReadBackWithClassNameLookup(new ModernLiteMessage(MODERN_PAYLOAD));

        assertThat(restored).isInstanceOf(ModernLiteMessage.class);
        assertThat(((ModernLiteMessage) restored).payload()).containsExactly(MODERN_PAYLOAD);
    }

    @Test
    void readResolveFallsBackToLegacyDefaultInstanceField() throws Exception {
        Object restored = serializeFormAndReadBackWithClassNameLookup(new LegacyLiteMessage(LEGACY_PAYLOAD));

        assertThat(restored).isInstanceOf(LegacyLiteMessage.class);
        assertThat(((LegacyLiteMessage) restored).payload()).containsExactly(LEGACY_PAYLOAD);
    }

    private static Object serializeFormAndReadBackWithClassNameLookup(MessageLite message) throws Exception {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        try (ClassNullingObjectOutputStream output = new ClassNullingObjectOutputStream(bytes)) {
            output.writeObject(SerializedFormAccess.serializedFormOf(message));
        }

        try (ObjectInputStream input = new ObjectInputStream(new ByteArrayInputStream(bytes.toByteArray()))) {
            return input.readObject();
        }
    }

    @SuppressWarnings("rawtypes")
    private abstract static class SerializedFormAccess extends GeneratedMessageLite {
        static Object serializedFormOf(MessageLite message) {
            return SerializedForm.of(message);
        }
    }

    private static class ClassNullingObjectOutputStream extends ObjectOutputStream {
        ClassNullingObjectOutputStream(OutputStream output) throws IOException {
            super(output);
            enableReplaceObject(true);
        }

        @Override
        protected Object replaceObject(Object object) throws IOException {
            if (object == ModernLiteMessage.class || object == LegacyLiteMessage.class) {
                return null;
            }
            return super.replaceObject(object);
        }
    }

    private abstract static class BaseLiteMessage implements MessageLite {
        private final byte[] payload;

        BaseLiteMessage(byte[] payload) {
            this.payload = Arrays.copyOf(payload, payload.length);
        }

        byte[] payload() {
            return Arrays.copyOf(payload, payload.length);
        }

        @Override
        public void writeTo(CodedOutputStream output) throws IOException {
            output.writeRawBytes(payload);
        }

        @Override
        public int getSerializedSize() {
            return payload.length;
        }

        @Override
        public Parser<? extends MessageLite> getParserForType() {
            throw new UnsupportedOperationException("Parsing is exercised through the builder");
        }

        @Override
        public ByteString toByteString() {
            return ByteString.copyFrom(payload);
        }

        @Override
        public byte[] toByteArray() {
            return payload();
        }

        @Override
        public void writeTo(OutputStream output) throws IOException {
            output.write(payload);
        }

        @Override
        public void writeDelimitedTo(OutputStream output) throws IOException {
            output.write(payload.length);
            output.write(payload);
        }

        @Override
        public boolean isInitialized() {
            return true;
        }
    }

    private static final class ModernLiteMessage extends BaseLiteMessage {
        @SuppressWarnings("unused")
        private static final ModernLiteMessage DEFAULT_INSTANCE = new ModernLiteMessage(new byte[0]);

        ModernLiteMessage(byte[] payload) {
            super(payload);
        }

        @Override
        public MessageLite.Builder newBuilderForType() {
            return new ModernBuilder();
        }

        @Override
        public MessageLite.Builder toBuilder() {
            return new ModernBuilder().mergeFrom(payload());
        }

        @Override
        public MessageLite getDefaultInstanceForType() {
            return DEFAULT_INSTANCE;
        }
    }

    private static final class LegacyLiteMessage extends BaseLiteMessage {
        @SuppressWarnings({"checkstyle:ConstantName", "unused"})
        private static final LegacyLiteMessage defaultInstance = new LegacyLiteMessage(new byte[0]);

        LegacyLiteMessage(byte[] payload) {
            super(payload);
        }

        @Override
        public MessageLite.Builder newBuilderForType() {
            return new LegacyBuilder();
        }

        @Override
        public MessageLite.Builder toBuilder() {
            return new LegacyBuilder().mergeFrom(payload());
        }

        @Override
        public MessageLite getDefaultInstanceForType() {
            return defaultInstance;
        }
    }

    private abstract static class BaseBuilder implements MessageLite.Builder {
        private byte[] payload = new byte[0];

        @Override
        public MessageLite.Builder clear() {
            payload = new byte[0];
            return this;
        }

        @Override
        public MessageLite build() {
            return buildPartial();
        }

        @Override
        public MessageLite.Builder clone() {
            BaseBuilder clone = newBuilder();
            clone.payload = Arrays.copyOf(payload, payload.length);
            return clone;
        }

        @Override
        public MessageLite.Builder mergeFrom(CodedInputStream input) throws IOException {
            return mergeFrom(input, ExtensionRegistryLite.getEmptyRegistry());
        }

        @Override
        public MessageLite.Builder mergeFrom(CodedInputStream input, ExtensionRegistryLite extensionRegistry)
                throws IOException {
            ByteArrayOutputStream bytes = new ByteArrayOutputStream();
            while (!input.isAtEnd()) {
                bytes.write(input.readRawByte());
            }
            return mergeFrom(bytes.toByteArray());
        }

        @Override
        public MessageLite.Builder mergeFrom(ByteString data) throws InvalidProtocolBufferException {
            return mergeFrom(data.toByteArray());
        }

        @Override
        public MessageLite.Builder mergeFrom(ByteString data, ExtensionRegistryLite extensionRegistry)
                throws InvalidProtocolBufferException {
            return mergeFrom(data);
        }

        @Override
        public MessageLite.Builder mergeFrom(byte[] data) {
            payload = Arrays.copyOf(data, data.length);
            return this;
        }

        @Override
        public MessageLite.Builder mergeFrom(byte[] data, int offset, int length) {
            payload = Arrays.copyOfRange(data, offset, offset + length);
            return this;
        }

        @Override
        public MessageLite.Builder mergeFrom(byte[] data, ExtensionRegistryLite extensionRegistry) {
            return mergeFrom(data);
        }

        @Override
        public MessageLite.Builder mergeFrom(byte[] data, int offset, int length,
                ExtensionRegistryLite extensionRegistry) {
            return mergeFrom(data, offset, length);
        }

        @Override
        public MessageLite.Builder mergeFrom(InputStream input) throws IOException {
            return mergeFrom(input.readAllBytes());
        }

        @Override
        public MessageLite.Builder mergeFrom(InputStream input, ExtensionRegistryLite extensionRegistry)
                throws IOException {
            return mergeFrom(input);
        }

        @Override
        public MessageLite.Builder mergeFrom(MessageLite message) {
            return mergeFrom(message.toByteArray());
        }

        @Override
        public boolean mergeDelimitedFrom(InputStream input) throws IOException {
            int length = input.read();
            if (length < 0) {
                return false;
            }
            return mergeFrom(input.readNBytes(length)) != null;
        }

        @Override
        public boolean mergeDelimitedFrom(InputStream input, ExtensionRegistryLite extensionRegistry)
                throws IOException {
            return mergeDelimitedFrom(input);
        }

        @Override
        public boolean isInitialized() {
            return true;
        }

        byte[] payload() {
            return Arrays.copyOf(payload, payload.length);
        }

        abstract BaseBuilder newBuilder();
    }

    private static final class ModernBuilder extends BaseBuilder {
        @Override
        public MessageLite buildPartial() {
            return new ModernLiteMessage(payload());
        }

        @Override
        public MessageLite getDefaultInstanceForType() {
            return ModernLiteMessage.DEFAULT_INSTANCE;
        }

        @Override
        BaseBuilder newBuilder() {
            return new ModernBuilder();
        }
    }

    private static final class LegacyBuilder extends BaseBuilder {
        @Override
        public MessageLite buildPartial() {
            return new LegacyLiteMessage(payload());
        }

        @Override
        public MessageLite getDefaultInstanceForType() {
            return LegacyLiteMessage.defaultInstance;
        }

        @Override
        BaseBuilder newBuilder() {
            return new LegacyBuilder();
        }
    }
}

File: tests/src/com.typesafe.akka/akka-protobuf-v3_3/2.9.0-M2/src/test/scala/com_typesafe_akka/akka_protobuf_v3_3/GeneratedMessageLiteInnerSerializedFormTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_typesafe_akka.akka_protobuf_v3_3

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.OutputStream
import java.nio.charset.StandardCharsets

import akka.protobufv3.internal.AbstractParser
import akka.protobufv3.internal.ByteString
import akka.protobufv3.internal.CodedInputStream
import akka.protobufv3.internal.CodedOutputStream
import akka.protobufv3.internal.ExtensionRegistryLite
import akka.protobufv3.internal.MessageLite
import akka.protobufv3.internal.Parser
import akka.protobufv3.internal.StringValue
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

import scala.annotation.static

class GeneratedMessageLiteInnerSerializedFormTest {
  @Test
  def readResolveUsesDefaultInstanceFieldFromLegacySerializedForm(): Unit = {
    val message: StringValue = StringValue.newBuilder().setValue("serialized-form").build()
    val serializedForm: Array[Byte] = legacySerializedForm(
      classOf[StringValue].getName,
      message.toByteArray
    )

    val deserialized: Object = deserialize(serializedForm)

    assertEquals(message, deserialized)
  }

  @Test
  def readResolveFallbackUsesLowerCaseDefaultInstanceFieldFromLegacySerializedForm(): Unit = {
    val messageClassName: String = classOf[FallbackNamedLiteMessage].getName
    val serializedForm: Array[Byte] = legacySerializedForm(messageClassName, Array.emptyByteArray)

    val deserialized: Object = deserialize(serializedForm)

    assertTrue(classOf[FallbackNamedLiteMessage].isInstance(deserialized))
    assertArrayEquals(
      Array.emptyByteArray,
      deserialized.asInstanceOf[FallbackNamedLiteMessage].toByteArray
    )
  }

  private def legacySerializedForm(messageClassName: String, asBytes: Array[Byte]): Array[Byte] = {
    val legacyForm = new LegacyGeneratedMessageLiteSerializedForm(asBytes, messageClassName)
    val originalBytes: Array[Byte] = serialize(legacyForm)
    replaceUtfValue(
      originalBytes,
      classOf[LegacyGeneratedMessageLiteSerializedForm].getName,
      "akka.protobufv3.internal.GeneratedMessageLite$SerializedForm"
    )
  }

  private def serialize(value: Object): Array[Byte] = {
    val bytes = new ByteArrayOutputStream()
    val stream = new ObjectOutputStream(bytes)
    try {
      stream.writeObject(value)
    } finally {
      stream.close()
    }
    bytes.toByteArray
  }

  private def deserialize(bytes: Array[Byte]): Object = {
    val stream = new ObjectInputStream(new ByteArrayInputStream(bytes))
    try {
      stream.readObject()
    } finally {
      stream.close()
    }
  }

  private def replaceUtfValue(bytes: Array[Byte], oldValue: String, newValue: String): Array[Byte] = {
    val oldBytes: Array[Byte] = oldValue.getBytes(StandardCharsets.UTF_8)
    val newBytes: Array[Byte] = newValue.getBytes(StandardCharsets.UTF_8)
    val oldToken: Array[Byte] = utfToken(oldBytes)
    val replacement: Array[Byte] = utfToken(newBytes)
    val index: Int = indexOf(bytes, oldToken)

    assertTrue(index >= 0, s"Serialized stream does not contain class descriptor for $oldValue")

    val result = new ByteArrayOutputStream()
    result.write(bytes, 0, index)
    result.write(replacement)
    result.write(bytes, index + oldToken.length, bytes.length - index - oldToken.length)
    result.toByteArray
  }

  private def utfToken(value: Array[Byte]): Array[Byte] = {
    assertTrue(value.length <= 0xffff)
    Array(((value.length >>> 8) & 0xff).toByte, (value.length & 0xff).toByte) ++ value
  }

  private def indexOf(bytes: Array[Byte], token: Array[Byte]): Int = {
    var index = 0
    while (index <= bytes.length - token.length) {
      var matched = true
      var tokenIndex = 0
      while (matched && tokenIndex < token.length) {
        matched = bytes(index + tokenIndex) == token(tokenIndex)
        tokenIndex += 1
      }
      if (matched) {
        return index
      }
      index += 1
    }
    -1
  }
}

@SerialVersionUID(0L)
final class LegacyGeneratedMessageLiteSerializedForm(
  private val asBytes: Array[Byte],
  private val messageClassName: String
) extends Serializable

final class FallbackNamedLiteMessage private (private val bytes: Array[Byte]) extends MessageLite {
  override def writeTo(output: CodedOutputStream): Unit = output.writeRawBytes(bytes)

  override def getSerializedSize(): Int = bytes.length

  override def getParserForType(): Parser[? <: MessageLite] = FallbackNamedLiteMessage.parser

  override def toByteString(): ByteString = ByteString.copyFrom(bytes)

  override def toByteArray(): Array[Byte] = bytes.clone()

  override def writeTo(output: OutputStream): Unit = output.write(bytes)

  override def writeDelimitedTo(output: OutputStream): Unit = {
    val codedOutput = CodedOutputStream.newInstance(output)
    codedOutput.writeUInt32NoTag(bytes.length)
    codedOutput.writeRawBytes(bytes)
    codedOutput.flush()
  }

  override def newBuilderForType(): MessageLite.Builder = new FallbackNamedLiteMessage.Builder()

  override def toBuilder(): MessageLite.Builder = new FallbackNamedLiteMessage.Builder().mergeFrom(bytes)

  override def getDefaultInstanceForType(): MessageLite = FallbackNamedLiteMessage.defaultInstance

  override def isInitialized(): Boolean = true
}

object FallbackNamedLiteMessage {
  @static val defaultInstance: FallbackNamedLiteMessage = new FallbackNamedLiteMessage(
    Array.emptyByteArray
  )

  private def parser: Parser[FallbackNamedLiteMessage] = new FallbackNamedLiteMessageParser(
    defaultInstance
  )

  final class Builder private[akka_protobuf_v3_3] () extends MessageLite.Builder {
    private var bytes: Array[Byte] = Array.emptyByteArray

    override def clear(): MessageLite.Builder = {
      bytes = Array.emptyByteArray
      this
    }

    override def build(): MessageLite = buildPartial()

    override def buildPartial(): MessageLite = new FallbackNamedLiteMessage(bytes.clone())

    override def clone(): MessageLite.Builder = new Builder().mergeFrom(bytes)

    override def mergeFrom(input: CodedInputStream): MessageLite.Builder = this

    override def mergeFrom(
      input: CodedInputStream,
      extensionRegistry: ExtensionRegistryLite
    ): MessageLite.Builder = this

    override def mergeFrom(byteString: ByteString): MessageLite.Builder = mergeFrom(
      byteString.toByteArray
    )

    override def mergeFrom(
      byteString: ByteString,
      extensionRegistry: ExtensionRegistryLite
    ): MessageLite.Builder = mergeFrom(byteString)

    override def mergeFrom(data: Array[Byte]): MessageLite.Builder = {
      bytes = data.clone()
      this
    }

    override def mergeFrom(
      data: Array[Byte],
      offset: Int,
      length: Int
    ): MessageLite.Builder = mergeFrom(data.slice(offset, offset + length))

    override def mergeFrom(
      data: Array[Byte],
      extensionRegistry: ExtensionRegistryLite
    ): MessageLite.Builder = mergeFrom(data)

    override def mergeFrom(
      data: Array[Byte],
      offset: Int,
      length: Int,
      extensionRegistry: ExtensionRegistryLite
    ): MessageLite.Builder = mergeFrom(data, offset, length)

    override def mergeFrom(input: InputStream): MessageLite.Builder = mergeFrom(
      input.readAllBytes()
    )

    override def mergeFrom(
      input: InputStream,
      extensionRegistry: ExtensionRegistryLite
    ): MessageLite.Builder = mergeFrom(input)

    override def mergeFrom(message: MessageLite): MessageLite.Builder = mergeFrom(message.toByteArray)

    override def mergeDelimitedFrom(input: InputStream): Boolean = {
      mergeFrom(input)
      true
    }

    override def mergeDelimitedFrom(
      input: InputStream,
      extensionRegistry: ExtensionRegistryLite
    ): Boolean = mergeDelimitedFrom(input)

    override def getDefaultInstanceForType(): MessageLite = FallbackNamedLiteMessage.defaultInstance

    override def isInitialized(): Boolean = true
  }
}

final class FallbackNamedLiteMessageParser(
  private val defaultInstance: FallbackNamedLiteMessage
) extends AbstractParser[FallbackNamedLiteMessage] {
  override def parsePartialFrom(
    input: CodedInputStream,
    extensionRegistry: ExtensionRegistryLite
  ): FallbackNamedLiteMessage = defaultInstance
}
